File: c++/DoublyCL.cpp
#include<iostream>
#include<stdlib.h>
using namespace std;


typedef struct node 
{
    int data;
    struct node *next;
    struct node *prev;
}NODE , *PNODE , *PPNODE;

class DoublyCL
{
    public:
        PNODE head;
        PNODE tail;
        int iCount;
        
    DoublyCL()
    {
        head = NULL;
        tail = NULL;
        iCount = 0;
    }
    void InsertFirst(int iNo)
    {
        PNODE newn = NULL;
        newn = new (NODE);
        newn->data = iNo;
        newn->next = NULL;
        newn->prev = NULL;

        if((head == NULL) && (tail == NULL))
        {
            head = newn;
            tail = newn;
        }
        else
        {
            newn->next = head;
            head = newn;
        }
        head->prev = tail;
        tail->next = head;
        iCount++;
    }
    void InsertLast(int iNo)
    {
        
        PNODE newn = NULL;
        newn = new (NODE);
        newn->data = iNo;
        newn->next = NULL;
        newn->prev = NULL;

        if((head == NULL) && (tail == NULL))
        {
            head = newn;
            tail = newn;
        }
        else
        {   
            tail->next = newn;
            newn->prev = tail;
            tail = newn;

        }
        head->prev = tail;
        tail->next = head;
        iCount++;
    }

    void InsertAtPos(int iNo , int iPos)
    {
        PNODE newn = NULL;
        PNODE temp = head;
        int i =0;

        if(iPos < 1 || iPos > iCount + 1)
        {
            cout<<"Invlaid Position :\n";
            return;
        }
        if(iPos == 1)
        {
            InsertFirst(iNo);

        }
        else if(iPos == iCount + 1)
        {
            InsertLast(iNo);
        }
        else 
        {
            newn = new (NODE);
            newn->data = iNo;
            newn->next = NULL;
            newn->prev = NULL;
            
            for(i = 1; i < iPos - 1; i++)
            {
                temp = temp->next;
            }
            newn->next = temp->next;
            temp->next->prev = newn;
            temp->next = newn;
            newn->prev = temp;

            iCount++;
        }
    }
    void DeleteFirst()
    {
        if((head == NULL ) && (tail  == NULL))
        {

            return;
        }
        else if(head == tail)
        {
            delete head;
            head = NULL;
            tail = NULL;
        }
        else 
        {
            head = head->next;
            delete tail->next;
        } 
        head->prev = tail;
        tail->next = head;
        iCount--;
    }

    void DeleteLast()
    {
        
        if((head == NULL ) && (tail  == NULL))
        {

            return;
        }
        else if(head == tail)
        {
            delete head;
            head = NULL;
            tail = NULL;
        }
        else 
        {
            tail = tail->prev;
            delete head->prev;

        }
        head->prev = tail;
        tail->next = head;
        iCount--;
    }

    void DeleteAtPos(int iPos)
    {
        int i = 0; 
        PNODE temp = head;
        PNODE target = NULL;

        if(iPos < 1 || iPos >  iCount)
        {
            cout<<"Invalid Position :\n";
            return;
        }
        if(iPos == 1)
        {
            DeleteFirst();
        }
        else if(iPos == iCount )
        {
            DeleteLast();
        }
        else 
        {
            for(i = 1; i < iPos-1; i++)
            {
                temp = temp->next;
            }
            target = temp->next;
            temp->next = temp->next->next;
            temp->next->prev = temp;
            delete target;            
        }        
    }
    void Display()
    {
        cout<<"Number  of Elements inside the Linked List are : \n";

        if((head == NULL) &&  (tail == NULL))
        {
            cout<<"Linked List is Empty : \n";
            return;
        }
        do 
        {
            cout<<"|"<<head->data<<"|<=>";
            head = head->next;
        }while(head != tail->next);
        cout<<"NULL\n";
    }

    int Count()
    {
        return iCount;
    }
};

int main()
{

    DoublyCL dobj;

    dobj.InsertFirst(101);
    
    dobj.InsertFirst(51);
    
    dobj.InsertFirst(21);

    dobj.InsertFirst(11);

    dobj.InsertLast(151);

    dobj.InsertAtPos(75,4);
    dobj.Display();

    dobj.DeleteFirst();
    dobj.DeleteLast();
    dobj.DeleteAtPos(2);

    dobj.Display();

    return 0;
}

File: Java/DoublyCL.java

class node
{
    public int data;
    public node next;
    public node prev;
}

class Doubly
{
    public node head;
    public node tail;
    public int iCount;

    public Doubly()
    {
        head = null;
        tail = null;
        iCount = 0;
    }
    public void Insertfirst(int ino)
    {   
        node newn = null;
        newn = new node();
        newn.data = ino;
        newn.next = null;
        newn.prev = null;
        if((head == null) && (tail == null))
        {
            head = newn;
            tail = newn;
        }
        else 
        {
            newn.next = head;
            head.prev = newn;
            head = newn;
        }
        head.prev = tail;
        tail.next = head;
        iCount++;
    }
    public void Insertlast(int iNo)
    {
        node newn = null;
        newn = new node();
        newn.data = iNo;
        newn.next = null;
        newn.prev = null;

        if((head == null) && (tail == null))
        {
            head = newn;
            tail = newn;
        }
        else 
        {
            tail.next = newn;
            newn.prev = tail;
            tail = newn;
        }
        head.prev = tail;
        tail.next = head;
        iCount++;          
    }
    public void InsertAtPos(int ino , int ipos)
    {
        node newn = null;
        node temp = null;

        if((ipos <  1)  || (ipos > iCount + 1))
        {
            System.out.println("Invalid Sysntax : ");
            return;
        }
        if(ipos == 1)
        {
            Insertfirst(ino);
        }
        else if(ipos == iCount + 1)
        {
            Insertlast(ino);
        }
        else 
        {
            int i = 0; 
            temp = head;
            newn = new node();
            newn.data = ino;
            newn.next = null;
            newn.prev = null;

            for(i = 1; i < ipos - 1; i++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next.prev = newn;
            newn.prev = temp;
            temp.next = newn;
            iCount++;
        }
    }

    public void DeleteFirst()
    {
        if((head == null ) &&  (tail == null))
        {
            System.out.println("Linked List is Empty ");
            return;
        }
        else if(head == tail)
        {
            head = null;
            tail = null;
            System.gc();
        }
        else
        {
            head = head.next;
            tail.next = null;
            System.gc();
        }
        head.prev = tail;
        tail.next = head;
        iCount--;
    }

    public void DeleteLast()
    {
        if((head == null ) &&  (tail == null))
        {
            System.out.println("Linked List is Empty ");
            return;
        }
        else if(head == tail)
        {
            head = null;
            tail = null;
            System.gc();
        }
        else
        {
            tail = tail.prev;
            tail.next = null;
            System.gc();
        }
        head.prev = tail;
        tail.next = head;   
        iCount--;
    }
    public void DeleteAtPos(int ipos)
    {
        node temp  = null;

        if((ipos < 1 ) || (ipos > iCount))
        {
            return;
        }
        if(ipos == 1)
        {
            DeleteFirst();
        }
        else if(ipos == iCount) 
        {
            DeleteLast();
        }
        else 
        {
            temp = head;
            int i = 0;
            for(i = 1; i < ipos - 1; i++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;
            temp.next.prev =temp;
            System.gc();
            iCount--;
        }
    }

    public void Display()
    {
        System.out.println("Number of Elements inside the Linked List are : ");
        node temp = head;

        do 
        {

            System.out.print("|"+temp.data+"|<=>");
            temp = temp.next;
        }while(temp != tail.next);  
        System.out.println("NULL");

    }

    public int Count()
    {
        return iCount; 
    }

}
class DoublyCL 
{
    public static void main(String a[])
    {
        Doubly dobj = new Doubly();

        dobj.Insertfirst(101);
        dobj.Insertfirst(51);
        dobj.Insertfirst(21);
        dobj.Insertfirst(11);
        dobj.Insertfirst(1);

        dobj.Insertlast(151);

        dobj.InsertAtPos(75, 5);
        
        dobj.Display();
        int iRet = dobj.Count();
        System.out.println("Number of Elements inside the Lineked list are : "+iRet);   
        
        dobj.DeleteFirst();
        dobj.DeleteLast();
        dobj.DeleteAtPos(3);
        dobj.Display();
        iRet = dobj.Count();
        System.out.println("Number of Elements inside the Lineked list are : "+iRet);   

    }
}